presto-common/src/main/java/com/facebook/presto/common/PrestoVersion.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoVersion
+        implements Comparable<PrestoVersion>
+{
+    private static final Pattern VERSION_REGEX = Pattern.compile("^(0\\.([1-9][0-9]*))(\\.([1-9][0-9]*))?(\\.*)?(-([2][0-9]{7}\\.[0-2][0-9]{5}-[0-9]{1,}|SNAPSHOT).*)?$");
+    private int majorVersion;
+    private int minorVersion;
+    private String snapshotVersion;
+    private String version;
+
+    public PrestoVersion(String version)
+    {
+        Matcher matcher = VERSION_REGEX.matcher(version);
+        this.majorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(2) : null).orElse("0"));
+        this.minorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(4) : null).orElse("0"));
+        this.snapshotVersion = Optional.ofNullable(matcher.lookingAt() ? matcher.group(7) : null).orElse("");
+        this.version = requireNonNull(version, "version string is null");
+    }
+
+    public String getVersion()
+    {
+        return version;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(this.majorVersion, this.minorVersion, this.snapshotVersion);
+    }
+
+    @Override
+    public boolean equals(Object version)
+    {
+        if (this == version) {
+            return true;
+        }
+        try {
+            PrestoVersion prestoVersion = (PrestoVersion) version;
+            return (Objects.equals(this.majorVersion, prestoVersion.majorVersion)
+                    && Objects.equals(this.minorVersion, prestoVersion.minorVersion)
+                    && (Objects.equals(this.snapshotVersion, prestoVersion.snapshotVersion)
+                    || Objects.equals(this.snapshotVersion, "SNAPSHOT")
+                        && !prestoVersion.snapshotVersion.isEmpty())
+                    || (Objects.equals(prestoVersion.snapshotVersion, "SNAPSHOT")
+                        && !this.snapshotVersion.isEmpty()));
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException("Please compare with a PrestoVersion object");
+        }
+    }
+
+    @Override
+    public int compareTo(PrestoVersion prestoVersion)
+    {
+        if (this.equals(prestoVersion)) {
```

Comment:
do you need the explicit call to equals to. If the control flow reaches the end it means that we have reached the point where all the fields are matching up and we can return 0.